File: apps/mobile/lib/features/tracking/data/gps_queue_repository.dart
import 'package:drift/drift.dart';
import 'package:uuid/uuid.dart';

import '../../../core/database/app_database.dart';

/// Écrit les positions relevées pendant une mission active dans la file
/// d'attente locale ([PendingGpsPositions]). N'appelle jamais l'API : seule
/// la couche de synchronisation (`SyncService`) parle au réseau (section 13
/// du cahier des charges — le flux GPS ne doit jamais bloquer sur le
/// réseau).
class GpsQueueRepository {
  GpsQueueRepository({required AppDatabase database}) : _db = database;

  final AppDatabase _db;
  final _uuid = const Uuid();

  Future<void> enqueue({
    required String vehicleId,
    String? missionId,
    required double latitude,
    required double longitude,
    double? accuracy,
    double? altitude,
    double? speed,
    double? heading,
    required bool isMocked,
    required DateTime recordedAt,
  }) {
    return _db.into(_db.pendingGpsPositions).insert(
          PendingGpsPositionsCompanion.insert(
            clientEventId: _uuid.v4(),
            vehicleId: vehicleId,
            missionId: Value(missionId),
            latitude: latitude,
            longitude: longitude,
            accuracy: Value(accuracy),
            altitude: Value(altitude),
            speed: Value(speed),
            heading: Value(heading),
            isMocked: Value(isMocked),
            recordedAt: recordedAt,
          ),
        );
  }

  /// Nombre de positions encore `pending`/`failed` (non confirmées par le
  /// serveur) — utilisé par la pastille de connectivité et l'écran de
  /// synchronisation.
  Stream<int> watchPendingCount() {
    final query = _db.selectOnly(_db.pendingGpsPositions)
      ..addColumns([_db.pendingGpsPositions.id.count()])
      ..where(
        _db.pendingGpsPositions.syncStatus.equalsValue(SyncStatus.pending) |
            _db.pendingGpsPositions.syncStatus.equalsValue(SyncStatus.failed),
      );
    return query.map((row) => row.read(_db.pendingGpsPositions.id.count()) ?? 0)
        .watchSingle();
  }

  /// Nombre de positions `failed` ayant épuisé leurs essais automatiques —
  /// affiché distinctement sur l'écran de synchronisation.
  Stream<int> watchFailedCount() {
    final query = _db.selectOnly(_db.pendingGpsPositions)
      ..addColumns([_db.pendingGpsPositions.id.count()])
      ..where(_db.pendingGpsPositions.syncStatus.equalsValue(SyncStatus.failed));
    return query
        .map((row) => row.read(_db.pendingGpsPositions.id.count()) ?? 0)
        .watchSingle();
  }

  /// Lot de positions prêtes à être envoyées : `pending`/`failed`, et
  /// respectant le backoff (`nextRetryAt`) sauf en synchronisation forcée
  /// (bouton manuel). Triées par [PendingGpsPositions.recordedAt] croissant.
  Future<List<PendingGpsPosition>> nextBatch({
    required int limit,
    required bool force,
    required int maxAutoRetries,
  }) {
    final now = DateTime.now();
    final query = _db.select(_db.pendingGpsPositions)
      ..where((t) {
        var predicate = t.syncStatus.equalsValue(SyncStatus.pending) |
            t.syncStatus.equalsValue(SyncStatus.failed);
        if (!force) {
          predicate = predicate & t.retryCount.isSmallerThanValue(maxAutoRetries);
          predicate = predicate &
              (t.nextRetryAt.isNull() | t.nextRetryAt.isSmallerOrEqualValue(now));
        }
        return predicate;
      })
      ..orderBy([(t) => OrderingTerm.asc(t.recordedAt)])
      ..limit(limit);
    return query.get();
  }

  Future<void> markUploading(List<int> ids) {
    return _db.batch((batch) {
      for (final id in ids) {
        batch.update(
          _db.pendingGpsPositions,
          const PendingGpsPositionsCompanion(
            syncStatus: Value(SyncStatus.uploading),
          ),
          where: (t) => t.id.equals(id),
        );
      }
    });
  }

  Future<void> markSynced(List<int> ids) {
    return _db.batch((batch) {
      for (final id in ids) {
        batch.update(
          _db.pendingGpsPositions,
          const PendingGpsPositionsCompanion(
            syncStatus: Value(SyncStatus.synced),
            lastError: Value(null),
            nextRetryAt: Value(null),
          ),
          where: (t) => t.id.equals(id),
        );
      }
    });
  }

  /// Marque une position en échec, incrémente `retryCount` et calcule le
  /// prochain essai autorisé avec un backoff exponentiel plafonné.
  Future<void> markFailed({
    required int id,
    required int previousRetryCount,
    String? error,
  }) {
    final retryCount = previousRetryCount + 1;
    return (_db.update(_db.pendingGpsPositions)..where((t) => t.id.equals(id)))
        .write(
      PendingGpsPositionsCompanion(
        syncStatus: const Value(SyncStatus.failed),
        retryCount: Value(retryCount),
        lastError: Value(error),
        nextRetryAt: Value(_nextRetryAt(retryCount)),
      ),
    );
  }

  /// Au démarrage de l'app : toute ligne restée `uploading` (process tué en
  /// plein envoi) est remise `pending` sans perte de données ni incrément
  /// de compteur d'essais.
  Future<void> resetStaleUploading() {
    return (_db.update(_db.pendingGpsPositions)
          ..where((t) => t.syncStatus.equalsValue(SyncStatus.uploading)))
        .write(
      const PendingGpsPositionsCompanion(syncStatus: Value(SyncStatus.pending)),
    );
  }
}

/// Backoff exponentiel : 5s, 10s, 20s, 40s... plafonné à 5 minutes.
DateTime _nextRetryAt(int retryCount) {
  final seconds = (5 * (1 << retryCount.clamp(0, 10))).clamp(5, 300);
  return DateTime.now().add(Duration(seconds: seconds));
}

File: apps/mobile/src/database/gps-queue.repository.ts
import { getDatabase } from './db';
import { PendingGpsPositionRow } from '../types/sync.types';
import { GpsPositionPayload } from '../types/tracking.types';

export interface EnqueueGpsParams {
  clientEventId?: string;
  vehicleId: string;
  missionId?: string | null;
  latitude: number;
  longitude: number;
  accuracy?: number | null;
  altitude?: number | null;
  speed?: number | null;
  heading?: number | null;
  isMocked?: boolean;
  recordedAt?: string;
}

export const GpsQueueRepository = {
  enqueue(params: EnqueueGpsParams): void {
    const db = getDatabase();
    const clientEventId = params.clientEventId || `gps_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
    const recordedAt = params.recordedAt || new Date().toISOString();
    const createdAtDevice = new Date().toISOString();

    db.runSync(
      `INSERT OR IGNORE INTO pending_gps_positions (
        client_event_id, vehicle_id, mission_id, latitude, longitude,
        accuracy, altitude, speed, heading, is_mocked,
        recorded_at, created_at_device, sync_status, retry_count
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'pending', 0)`,
      [
        clientEventId,
        params.vehicleId,
        params.missionId || null,
        params.latitude,
        params.longitude,
        params.accuracy ?? null,
        params.altitude ?? null,
        params.speed ?? null,
        params.heading ?? null,
        params.isMocked ? 1 : 0,
        recordedAt,
        createdAtDevice,
      ]
    );
  },

  getNextBatch(limit = 100, force = false, maxAutoRetries = 8): PendingGpsPositionRow[] {
    const db = getDatabase();
    const nowIso = new Date().toISOString();

    if (force) {
      return db.getAllSync<PendingGpsPositionRow>(
        `SELECT * FROM pending_gps_positions 
         WHERE sync_status != 'synced' AND sync_status != 'uploading'
         ORDER BY recorded_at ASC LIMIT ?`,
        [limit]
      );
    }

    return db.getAllSync<PendingGpsPositionRow>(
      `SELECT * FROM pending_gps_positions 
       WHERE sync_status IN ('pending', 'failed')
         AND retry_count < ?
         AND (next_retry_at IS NULL OR next_retry_at <= ?)
       ORDER BY recorded_at ASC LIMIT ?`,
      [maxAutoRetries, nowIso, limit]
    );
  },

  markUploading(ids: number[]): void {
    if (ids.length === 0) return;
    const db = getDatabase();
    const placeholders = ids.map(() => '?').join(',');
    db.runSync(
      `UPDATE pending_gps_positions SET sync_status = 'uploading' WHERE id IN (${placeholders})`,
      ids
    );
  },

  markSynced(ids: number[]): void {
    if (ids.length === 0) return;
    const db = getDatabase();
    const placeholders = ids.map(() => '?').join(',');
    db.runSync(
      `DELETE FROM pending_gps_positions WHERE id IN (${placeholders})`,
      ids
    );
  },

  markFailed(id: number, previousRetryCount: number, error: string): void {
    const db = getDatabase();
    const nextRetryCount = previousRetryCount + 1;
    // Exponential backoff: 5s * 2^(retryCount - 1), capped at 5 minutes
    const delaySeconds = Math.min(300, 5 * Math.pow(2, Math.max(0, nextRetryCount - 1)));
    const nextRetryAt = new Date(Date.now() + delaySeconds * 1000).toISOString();

    db.runSync(
      `UPDATE pending_gps_positions 
       SET sync_status = 'failed', retry_count = ?, last_error = ?, next_retry_at = ?
       WHERE id = ?`,
      [nextRetryCount, error, nextRetryAt, id]
    );
  },

  resetStaleUploading(): void {
    const db = getDatabase();
    db.runSync(`UPDATE pending_gps_positions SET sync_status = 'pending' WHERE sync_status = 'uploading'`);
  },

  getPendingCount(): number {
    const db = getDatabase();
    const row = db.getFirstSync<{ count: number }>(
      `SELECT COUNT(*) as count FROM pending_gps_positions WHERE sync_status != 'synced'`
    );
    return row?.count || 0;
  },

  getFailedCount(): number {
    const db = getDatabase();
    const row = db.getFirstSync<{ count: number }>(
      `SELECT COUNT(*) as count FROM pending_gps_positions WHERE sync_status = 'failed'`
    );
    return row?.count || 0;
  },

  toPayload(row: PendingGpsPositionRow): GpsPositionPayload {
    return {
      clientEventId: row.client_event_id,
      vehicleId: row.vehicle_id,
      missionId: row.mission_id || null,
      latitude: row.latitude,
      longitude: row.longitude,
      accuracy: row.accuracy,
      altitude: row.altitude,
      speed: row.speed,
      heading: row.heading,
      isMocked: row.is_mocked === 1,
      recordedAt: row.recorded_at,
    };
  },
};
